src/manager/checkpoint/manager/bottomup.rs

```diff
@@ -1,49 +1,288 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use crate::config::Subnet;
+use crate::lotus::client::DefaultLotusJsonRPCClient;
+use crate::lotus::message::mpool::MpoolPushMessage;
+use crate::lotus::LotusClient;
 use crate::manager::checkpoint::CheckpointManager;
+use anyhow::anyhow;
 use async_trait::async_trait;
+use cid::Cid;
+use fil_actors_runtime::cbor;
 use fvm_shared::address::Address;
 use fvm_shared::clock::ChainEpoch;
+use fvm_shared::MethodNum;
 use ipc_gateway::BottomUpCheckpoint;
 use ipc_sdk::subnet_id::SubnetID;
+use primitives::TCid;
+use std::str::FromStr;
 
-pub struct BottomUpCheckpointManager;
+pub struct BottomUpCheckpointManager {
+    parent: SubnetID,
+    parent_client: DefaultLotusJsonRPCClient,
+    child_subnet: Subnet,
+    child_client: DefaultLotusJsonRPCClient,
+
+    checkpoint_period: ChainEpoch,
+}
+
+impl BottomUpCheckpointManager {
+    pub fn new_with_period(
+        parent_subnet: SubnetID,
+        parent_client: DefaultLotusJsonRPCClient,
+        child_subnet: Subnet,
+        child_client: DefaultLotusJsonRPCClient,
+        checkpoint_period: ChainEpoch,
+    ) -> Self {
+        Self {
+            parent: parent_subnet,
+            parent_client,
+            child_subnet,
+            child_client,
+            checkpoint_period,
+        }
+    }
+
+    pub async fn new(
+        parent_client: DefaultLotusJsonRPCClient,
+        parent: SubnetID,
+        child_client: DefaultLotusJsonRPCClient,
+        child_subnet: Subnet,
+    ) -> anyhow::Result<Self> {
+        let checkpoint_period = obtain_checkpoint_period(&child_subnet.id, &parent_client).await?;
+        Ok(Self::new_with_period(
+            parent,
+            parent_client,
+            child_subnet,
+            child_client,
+            checkpoint_period,
+        ))
+    }
+
+    async fn last_executed_epoch(&self) -> anyhow::Result<ChainEpoch> {
+        let parent_head = self.parent_client.chain_head().await?;
+
+        // A key assumption we make now is that each block has exactly one tip set. We panic
+        // if this is not the case as it violates our assumption.
+        // TODO: update this logic once the assumption changes (i.e., mainnet)
+        assert_eq!(parent_head.cids.len(), 1);
+
+        let cid_map = parent_head.cids.first().unwrap().clone();
+        let parent_tip_set = Cid::try_from(cid_map)?;
```

Comment:
Should we make the `parent_tipset` an argument for this one? That way we just need to get it one and we can reuse it in the rest of the functions.

Reply:
I have placed the obtain validators into a common function and taken it out of the trait. Also placed the latest executed method into the trait so that all validators can share the same parameter during each epoch submission.